stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -414,6 +414,84 @@ function triu(A::Symmetric, k::Integer=0)
     end
 end
 
+function dot(A::Symmetric, B::Symmetric)
+    n = size(A, 2)
+    if n != size(B, 2)
+        throw(DimensionMismatch("A has dimensions $(size(A)) but B has dimensions $(size(B))"))
+    end
+
+    dotprod = zero(dot(first(A), first(B)))
+    @inbounds if A.uplo == 'U' && B.uplo == 'U'
+        for j in 1:n
+            for i in 1:(j - 1)
+                dotprod += 2 * dot(A.data[i, j], B.data[i, j])
+            end
+            dotprod += dot(A[j, j], B[j, j])
+        end
+    elseif A.uplo == 'L' && B.uplo == 'L'
+        for j in 1:n
+            dotprod += dot(A[j, j], B[j, j])
+            for i in (j + 1):n
+                dotprod += 2 * dot(A.data[i, j], B.data[i, j])
+            end
+        end
+    elseif A.uplo == 'U' && B.uplo == 'L'
+        for j in 1:n
+            for i in 1:(j - 1)
+                dotprod += 2 * dot(A.data[i, j], transpose(B.data[j, i]))
+            end
+            dotprod += dot(A[j, j], B[j, j])
+        end
+    elseif A.uplo == 'L' && B.uplo == 'U'
```

Comment:
This is the last remaining combination, so this could be shortened to simply `else` without condition. Similarly in the Hermitian case.